File: packages/ruby-sdk/lib/strait/authoring/event.rb
# frozen_string_literal: true

module Strait
  module Authoring
    # An event definition with key and optional validator.
    class EventDefinition
      attr_reader :key

      def initialize(key, validate: nil)
        @key = key
        @validate = validate
      end

      def parse(input)
        @validate ? @validate.call(input) : input
      end
    end

    # Creates a new event definition.
    def self.define_event(key, validate: nil)
      EventDefinition.new(key, validate: validate)
    end
  end
end

File: packages/go-sdk/authoring/event.go
package authoring

// EventDefinition represents a named event with optional validation.
type EventDefinition struct {
	Key      string
	Validate func(input any) (any, error)
}

// Parse validates and returns the input using the validator, or returns as-is.
func (e *EventDefinition) Parse(input any) (any, error) {
	if e.Validate != nil {
		return e.Validate(input)
	}
	return input, nil
}

// DefineEvent creates a new event definition.
func DefineEvent(key string, validate func(any) (any, error)) *EventDefinition {
	return &EventDefinition{Key: key, Validate: validate}
}
